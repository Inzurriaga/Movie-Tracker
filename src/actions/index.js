// what gets sent
export const getMovies = (movies) => ({
  type: 'GET_MOVIES',
  movies
});

export const addUser = (userInfo) => ({
  type: "UPDATE_USER",
  userInfo
})

export const getGenres = (genre) => ({
  type: "GET_GENRES",
  genre
})
