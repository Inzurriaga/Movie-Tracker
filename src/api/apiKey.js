export const apiKey = '3033b069b2d0b32039c7d85cfe10cfd3';
