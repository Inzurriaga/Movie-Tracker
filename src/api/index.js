import { apiKey } from './apiKey'

const backEndUrl = 'http://localhost:3000/api/'

// Movies - DB Fetch Calls
const fetchMovies = async (url, urlEnd) => {
  let movieUrl = `${url}${apiKey}${urlEnd}`;
    const response = await fetch(movieUrl);
    const movies = await response.json()
    return movies
}

// Users - Backend Fetch Calls
const settings = (method, body) => ({
  method,
  headers: {'Content-Type': 'application/json'},
  body: JSON.stringify(body)
})

const deleteSettings = () => ({
  method: 'DELETE',
  headers: {'Content-Type': 'application/json'}
})

const postFetch = async (url, method, body) => {
  const response = await fetch(`${backEndUrl}${url}`, settings(method, body))
  return response;
}

const getFetch = async (url) => {
  const response = await fetch(`${backEndUrl}${url}`)
  if(!response.ok) {
    return 'User Already Exists'
    // throw Error(response.statusText)
  } else {
    const user = await response.json()
    return user.data;
  }
    // if(response.status === 'success') {}
}

const deleteFetch = async (url) => {
  const response = await fetch(`${backEndUrl}${url}`, deleteSettings())
  const retrievedData = await response.json()
  return retrievedData
}





export { fetchMovies, postFetch, getFetch, deleteFetch, settings, deleteSettings }
