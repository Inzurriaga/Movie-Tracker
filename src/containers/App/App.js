import React, { Component } from 'react';
import { Route, Switch } from "react-router-dom"
import { connect } from "react-redux"
import { getDiscover, getGenres } from '../../actions';
import { fetchMovies } from "../../api/index";
import MainPage from '../MainPage/MainPage'
import MovieInfo from "../MovieInfo/MovieInfo"
import Header from '../../containers/Header/Header'
import Footer from '../../components/Footer/Footer'
import Movies from "../Movies/Movies"

export class App extends Component {
  constructor() {
    super();
    this.state = {
      moviesCat: [
        "&sort_by=popularity.desc",
        "&with_genres=28",
        "&with_genres=27",
        "&with_genres=35"
      ],
      urlMovies: "https:api.themoviedb.org/3/discover/movie?"
    }
  }

  componentDidMount = async () => {
    const { moviesCat, urlMovies } = this.state;
    try {
      const unresolvedAllMovies = moviesCat.map( async (movie) => {
        const singleCategory = await fetchMovies(urlMovies, movie)
        return singleCategory
      })
      const allMovies = await Promise.all(unresolvedAllMovies)
      this.props.getDiscover(allMovies[0].results)
      this.setGenres(allMovies)
    } catch(error) {
      console.log(error.message)
    }
  }

  setGenres = (response) => {
    const genres = response.filter((genre, index) => index > 0 )
    this.props.getGenres(genres)
  }

  render() {
    return (
      <div className="App">
        <Header />
        <Switch>
          <Route exact path="/" component={MainPage} />
          {
            this.props.genres.length &&
            <Route exact path="/movies/genre/:id" render={({match}) => {
              const { id } = match.params
              const genre = this.props.genres.find((genre, index) => index === parseInt(id))
              return <Movies id="genres" genreInfo={genre}/>
            }} />
          }
          <Route exact path="/movies/allMovies" render={() => {
            return <Movies key="allMovies" id="allMovies" />
          }} />
          <Route exact path="/favorites" render={() => {
            return <Movies key="favorites" id="favorites" />
          }} />
          <Route exact path="/InTheater" render={() => {
            return <Movies key="favorites" id="inTheater" />
          }} /> 
           <Route exact path="/upcoming" render={() => {
            return <Movies key="upcoming" id="upcoming" />
          }} /> 
          {
            this.props.movies.length &&
            <Route exact path="/movies/:id" render={({match}) => {
              const { id } = match.params
              return <MovieInfo movieID={id}/>
            }} />
          }
        </Switch>
        <Footer />
      </div>
    );
  }
}

export const mapStateToProps = (state) => ({
  movies: state.movies,
  genres: state.genres
})

export const mapDispatchToProps = (dispatch) => ({
  getDiscover: (movies) => dispatch(getDiscover(movies)),
  getGenres: (genre) => dispatch(getGenres(genre)),
})

export default connect(mapStateToProps, mapDispatchToProps)(App)
