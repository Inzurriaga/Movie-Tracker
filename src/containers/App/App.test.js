import React from 'react';
import { shallow } from "enzyme"
import { App, mapDispatchToProps, mapStateToProps  } from './App';
import { fetchMovies } from "../../api/index"
import { getDiscover, getGenres } from '../../actions';
jest.mock("../../api/index")

const mockGenres = [{namne:"hello"},{name: "what"},{name: "yes"}]
const mockMovies = [{namne:"hello"},{name: "what"},{name: "yes"}]
const mockUrl = "https:api.themoviedb.org/3/discover/movie?"

describe('app', () => {
    describe("app", () => {
        let wrapper; 
        beforeEach(() => {
            wrapper = shallow(
                <App genres={mockGenres}
                    movies={mockMovies}/>
            )
        })
        it("should have a snap shot", () => {
            expect(wrapper).toMatchSnapshot()
        });
        it("should have a default state", () => {
            expect(wrapper.state()).toEqual({
                moviesCat: [
                  "&sort_by=popularity.desc",
                  "&with_genres=28",
                  "&with_genres=27",
                  "&with_genres=35"
                ],
                urlMovies: "https:api.themoviedb.org/3/discover/movie?",
                error: ""
              })
        });
        it("should invoke the fetch call on component did mount", () => {

        });
        it("fetch call takes expoect url", () => {
            wrapper.instance().componentDidMount()
            expect(fetchMovies).toHaveBeenCalledWith(mockUrl, "&with_genres=28")
        })

    })
    describe("mapStateToProps", () => {
        it("should return a object", () => {
            const mockData = {
                movies: [],
                genres: [],
                category: []
            }
            const expected = {
                movies: [],
                genres: []
            }
            const mockprops = mapStateToProps(mockData)
            expect(mockprops).toEqual(expected)
        })
    })
    describe("mapDispatchToProps", () => {
        it("should call dispach for getDiscover", () => {
            const mockDispatch = jest.fn()
            const actionToDispatch = getDiscover(mockMovies)
            const mappedProps = mapDispatchToProps(mockDispatch)
            mappedProps.getDiscover(mockMovies)
            expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
        });
        it("should call dispach for getGenres", () => {
            const mockDispatch = jest.fn()
            const actionToDispatch = getGenres(mockMovies)
            const mappedProps = mapDispatchToProps(mockDispatch)
            mappedProps.getGenres(mockMovies)
            expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
        })
    })
});
