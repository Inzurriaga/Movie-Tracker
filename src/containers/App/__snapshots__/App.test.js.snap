// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app app should have a snap shot 1`] = `
<div
  className="App"
>
  <Connect(Header) />
  <Switch>
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
    <Route
      exact={true}
      path="/movies/:title/:id"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/movies/allMovies"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/favorites"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/InTheater"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/upcoming"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/movies/:id"
      render={[Function]}
    />
  </Switch>
</div>
`;
